SEED-platform/seed
# Matching update

**Note/Reminder**:This PR includes changes to pre-existing migrations. Deployments that had migrations after `0104_auto_20190509_1854` will need to have any proceeding migrations rolled back, then the migrations within this PR will need to be applied. Presumably, this only made it as far `dev1` - further considerations would likely be needed if other deployments are affected.

#### Any background context you want to provide?
In order to more easily implement year-over-year functionality, it was decided to revamp matching. 

The purpose of these changes were to simplify the matching logic while giving users the power to decide what constitutes a match. Along the way, a pre-emptive, though still related, update was made to merge roll-up matches on different occasions.

Note: Technically, any of the merge roll-ups are reversible.

#### What's this PR do?
The following changes have been made:
- User can specify matching criteria columns with `is_matching_criteria` field on Column model.
- Extra data columns can't be matching criteria.
- There must be at least one matching criteria.
- The mapping page during file import lists these matching criteria columns in the texts and warnings.
- heavy refactor of record **import** matching logic to be more DB dependent, (hopefully) improve transparency, and reusability
- Using code mentioned in bullet above, match-merge logic created for a single, **existing** record sets (-States, -Views, Canonical Record).
- Match-merge round run post record-edit - roll ups can occur here.
- Match-merge round run post records-merge - roll ups can occur here.
- Match-merge round run on incoming records - roll ups can occur here. (This existed before but only the 1st of many matches was merged.)
- During any merge roll ups, priority is given to most recently updated records (by corresponding AuditLog dates), then, if applicable, the record in focus (on edit or merge).
- Merging should still combine/preserve labels, notes, pairings, and if applicable, meters.
- (Pre-emptive update) Merging and unmerging records creates new canonical IDs, while only preserving the old ones if they're used in other -Views.

#### How should this be manually tested?
With a dataset of properties and tax lots having labels, notes, pairings, and meters, take the following actions in a variety of orders. While doing this, make note of any areas that could use more/less warning or feedback for users.
- update the matching criteria via the Column Settings page
- import matching properties and tax lots ("matching" according to your specified criteria)
- manually merge properties and tax lots
    - merges should result in roll ups if there are records that match the resulting record
- via the inventory detail page, edit a record to match or that already matches other record(s) and ensure roll ups occur.
- Unmerging any records should be the same (even though canonical records are usually changed).
